src/common/ad_tx.c

```diff
@@ -608,3 +613,423 @@ ad_tx_redo_act_next(struct ad_tx *tx)
 	}
 	return &tx->tx_redo_act_pos->it_act;
 }
+
+static struct ad_tx *
+tx_get()
+{
+	return tls_tx;
+}
+
+static void
+tx_set(struct ad_tx *tx)
+{
+	tls_tx = tx;
+}
+
+static void
+tx_callback(struct ad_tx *tx)
+{
+	if (!tx->tx_stage_cb || tx->tx_layer != 0)
+		return;
+	tx->tx_stage_cb(ad_tx_stage(tx), tx->tx_stage_cb_arg);
+}
+
+static int
+tx_end(struct ad_tx *tx, int err)
+{
+	struct umem_tx	*utx;
+	int		 rc;
+
+	if (err == 0)
+		err = tx->tx_last_errno;
+	else
+		tx->tx_last_errno = err;
+
+	D_ASSERTF(tx->tx_layer >= 0, "TX "DF_U64", bad layer %d\n", ad_tx_id(tx), tx->tx_layer);
+	if (tx->tx_layer != 0)
+		return 0;
+
+	rc = ad_tx_end(tx, err);
+	if (err == 0) {
+		if (rc == 0) {
+			ad_tx_stage_set(tx, UMEM_STAGE_ONCOMMIT);
+		} else {
+			D_ERROR("ad_tx_end(%d) failed, "DF_RC"\n", err, DP_RC(rc));
+			tx->tx_last_errno = rc;
+			ad_tx_stage_set(tx, UMEM_STAGE_ONABORT);
+		}
+	} else {
+		D_ASSERT(rc != 0);
+		D_ASSERTF(ad_tx_stage(tx) == UMEM_STAGE_ONABORT,
+			  "TX "DF_U64", bad stage %d\n", ad_tx_id(tx), ad_tx_stage(tx));
+	}
+
+	tx_callback(tx);
+	if (tx_get() == tx)
+		tx_set(NULL);
+	/* Give a chance to UMEM_STAGE_NONE callback.
+	 * XXX with less opportunity comparing with PMDK as user may started new TX here,
+	 *     or change to always call it even if with new TX in TLS?
+	 */
+	if (tx_get() == NULL) {
+		ad_tx_stage_set(tx, UMEM_STAGE_NONE);
+		tx_callback(tx);
+	}
+	rc = tx->tx_last_errno;
+	utx = ad_tx2umem_tx(tx);
+	D_FREE(utx);
+
+	return rc;
+}
+
+static int
+tx_abort(struct ad_tx *tx, int err)
+{
+	if (err == 0)
+		err = -DER_CANCELED;
+
+	ad_tx_stage_set(tx, UMEM_STAGE_ONABORT);
+
+	return tx_end(tx, err);
+}
+
+int
+mo_ad_tx_begin(struct umem_instance *umm, struct umem_tx_stage_data *txd)
+{
+	struct umem_tx		*utx = NULL;
+	struct ad_tx		*tx;
+	struct ad_blob_handle	 bh = umm2ad_blob_hdl(umm);
+	struct ad_blob		*blob;
+	struct umem_store	*store;
+	uint64_t		 tx_id;
+	int			 rc = 0;
+
+	blob = bh.bh_blob;
+	tx = tx_get();
+	if (tx == NULL) {
+		D_ALLOC_PTR(utx);
+		if (utx == NULL)
+			return -DER_NOMEM;
+
+		tx = umem_tx2ad_tx(utx);
+		rc = ad_tx_begin(bh, tx);
+		if (rc) {
+			D_ERROR("ad_tx_begin failed, "DF_RC"\n", DP_RC(rc));
+			D_FREE(utx);
+			return rc;
+		}
+
+		store = &blob->bb_store;
+		rc = store->stor_ops->so_wal_reserv(store, &tx_id);
+		if (rc) {
+			D_ERROR("so_wal_reserv failed, "DF_RC"\n", DP_RC(rc));
+			blob_decref(blob); /* drop ref taken in ad_tx_begin */
+			D_FREE(utx);
+			return rc;
+		}
+
+		/* possibly yield in so_wal_reserv, but tls_tx should be NULL when it get back */
+		D_ASSERT(tx_get() == NULL);
+		tx->tx_stage_cb = umem_stage_callback;
+		tx->tx_stage_cb_arg = txd;
+		ad_tx_id_set(tx, tx_id);
+		ad_tx_stage_set(tx, UMEM_STAGE_WORK);
+		tx_set(tx);
+		D_DEBUG(DB_TRACE, "TX "DF_U64" started\n", tx_id);
+	} else if (ad_tx_stage(tx) == UMEM_STAGE_WORK) {
```

Comment:
As long as you get 'tx' from TLS, it must be a nested transaction, right? I suppose we can assert tx stage == WORK here?

Reply:
Ok, to me, we could regard it as a bug and assert tx_stage is WORK. I'm fine with returning an error too, but this error shouldn't override the original error code.

Reply:
OK, let's assert it then, it is a little simpler. I'll change it if need refresh this PR or later. thx